### DISCUSSION TITLE
How to define Composer auth tokens in repository env variables?

### DISCUSSION THREAD

**Discussion by aeno** [renovatebot/renovate#27077] [Request Help] [answered]
### How are you running Renovate?

Self-hosted

### If you're self-hosting Renovate, tell us what version of Renovate you run.

37.31.2

### If you're self-hosting Renovate, select which platform you are using.

GitLab self-hosted

### What is your question?

We set up Renovate as a repository in out self-hosted GitLab instance and try to run it against multiple project repositories. Each project depends on external Composer packages that can only be fetched with different Composer repo auth tokens.

We don't want to keep a large list of `hostRule` tokens in the `config.js` bot config and neither want these tokens to be committed to the project's `renovate.json` file.

Is there any possibility to provide `hostRule`s on a per-project basis via GitLab CI variables or the like?

**Comment by viceice** [comment, marked as answer]
no, but you can use config encryption (gpg) to add encrypted token to repo config securely. you need to keep the private key secure of cause.

**Reply by aeno** [reply]
Thanks, I think we will stick to that. :+1:

**Comment by rarkins** [comment]
hostRules can be defined using env, e.g. RENOVATE_HOST_RULES or put them all in RENOVATE_CONFIG
